crates/matrix-sdk/src/event_cache/mod.rs

```diff
@@ -396,6 +404,49 @@ impl EventCacheInner {
                 continue;
             };
 
+            // Handle and cache events and relations
+            let mut relations = self.relations.write().await;
+            for ev in &joined_room_update.timeline.events {
+                if let Ok(AnySyncTimelineEvent::MessageLike(ev)) = ev.event.deserialize() {
+                    if let Some(AnyMessageLikeEventContent::RoomMessage(c)) = ev.original_content()
+                    {
+                        if let Some(Relation::Replacement(relation)) = c.relates_to {
+                            if let Some(replacements) = relations.get_mut(&relation.event_id) {
+                                replacements.insert(ev.event_id().to_owned());
+                            } else {
+                                relations.insert(
+                                    relation.event_id,
+                                    BTreeSet::from_iter(vec![ev.event_id().to_owned()]),
+                                );
+                            }
```

Comment:
You can probably use `relations.entry(relation.event_id.clone()).or_default().insert(…)` to make this a bit shorter to read.